base/io.jl

```diff
@@ -475,10 +484,39 @@ read(filename::AbstractString, args...) = open(io->read(io, args...), convert(St
 read(filename::AbstractString, ::Type{T}) where {T} = open(io->read(io, T), convert(String, filename)::String)
 
 """
-    read!(stream::IO, array::AbstractArray)
-    read!(filename::AbstractString, array::AbstractArray)
+    read!(stream::IO, a::AbstractArray{T})
+    read!(stream::IO, a::Ref{T})
+    read!(filename::AbstractString, a::Union{AbstractArray{T}, Ref{T}})
+
+Read `length(a)` consecutive values of type `T` from an I/O stream or
+file into `a`, returning `a`.
+
+If `a` is an `Array` and `isbitstype(T)` is true, this function reads
+`size(a)` bytes directly into memory. Otherwise, each element of
+`a::AbstractArray{T}` is read with a call to `read(stream, T)`.
+
+The `Ref{T}` variant is not intended for use if `isbitstype(T)` is
+false.
```

Comment:
This is seems broken to me.  It really seems like we should have a safe fallback:
```jl
function read!(s::IO, x::Ref{T}) where {T}
    if isbitstype(T)
        unsafe_read(s, x, Core.sizeof(T))
    else
        x[] = read(s, T)
    end
    return x
end
```

Reply:
@stevengj I think I agree. Since a `Ref{T}` is essentially a 0-dimensional array of length 1, `read!` should work the same for `Ref` and `Array`. That then will also simplify its docstring.

Reply:
Thanks for pointing that out. I made a PR to fix that https://github.com/JuliaLang/julia/pull/51949